### ISSUE TITLE
CAKE kernel candidates for MiniMax-H3 on B300 and RTX 5090

### ISSUE THREAD

**Issue by yyihuang** [flashinfer-ai/flashinfer#4532]
# CAKE kernel candidates for MiniMax-H3 on B300 and RTX 5090

Thank you for inviting customized kernel challenges in the CAKE tracker. We have
a MiniMax-H3 FL2VA runtime based on Sol Engine and SGLang and would like to ask
whether any of the standalone kernels below would be useful CAKE challenges.

These are independent candidates rather than a request for the CAKE team to
implement all of them. We will take care of Sol Engine/SGLang integration,
collectives, quantized checkpoints, calibration, end-to-end validation, and
deployment.

The two targets are:

- NVIDIA B300 / SM103a, with Ulysses SP1/SP2/SP4/SP8;
- NVIDIA RTX 5090 / SM120, quantized execution and SP1 only.

## B300 / SM103a candidates

Common H3 dimensions:

- hidden size: 5,376
- heads: 56
- head dimension: 128
- attention inner dimension: 7,168
- FFN dimension: 14,336
- supported Ulysses degree `P`: 1, 2, 4, or 8
- heads per rank after all-to-all: `56 / P` = 56, 28, 14, or 7
- non-causal packed-varlen attention
- batch size: 1

For B300, we are interested in separate shape specializations for SP1, SP2,
SP4, and SP8.

Nominal `M` values for our main 768p duration buckets are:

| Duration | Global packed rows `T` | SP1 `M` | SP2 `M` | SP4 `M` | SP8 `M` |
|---:|---:|---:|---:|---:|---:|
| 4 s | 33,472 | 33,472 | 16,736 | 8,368 | 4,184 |
| 5 s | 38,592 | 38,592 | 19,296 | 9,648 | 4,824 |
| 6 s | 48,768 | 48,768 | 24,384 | 12,192 | 6,096 |
| 8 s | 58,944 | 58,944 | 29,472 | 14,736 | 7,368 |
| 10 s | 74,240 | 74,240 | 37,120 | 18,560 | 9,280 |
| 15 s | 109,952 | 109,952 | 54,976 | 27,488 | 13,744 |

Prompt length makes `M` slightly dynamic, so kernels need tail handling around
these values. We can provide exact tensor traces and a PyTorch reference for each
operator below.

## 1. Fused BF16 pre-attention kernel — highest priority

Reference operation:

```text
x [M, 5376]
  -> RMSNorm
  -> indexed AdaLN scale/shift
  -> BF16 QKV GEMM [M, 5376] x [5376, 21504]
  -> Q/K RMSNorm over D=128
  -> 3-D RoPE on Q/K
  -> destination-major BF16 QKV
       [P, M, 56/P, 3, 128]
```

The output is only a send buffer; NCCL all-to-all is outside the kernel. The goal
is to avoid writing normalized x, modulated x, ordinary QKV, normalized Q/K, and
RoPE Q/K to HBM. `P` may be implemented as four compile-time specializations
instead of one dynamic kernel if that produces better code.

## 2. Fused low-precision pre-attention kernels

Same operation and output layout as kernel 1, with two independent variants:

```text
A. MXFP8/FP8 QKV GEMM
   -> BF16 Q/K RMSNorm + RoPE
   -> destination-major E4M3 QKV + scales

B. NVFP4 W4A4 QKV GEMM
   -> BF16 Q/K RMSNorm + RoPE
   -> destination-major NVFP4 QKV + block scales
```

An additional NVFP4-GEMM-to-E4M3-output variant is useful if the matching NVFP4
attention kernel is not accurate enough.

We will provide prepacked weights and scale tensors. The kernel does not need to
create or calibrate quantized weights.

### Candidate 2 implementation tracking

- [ ] https://github.com/flashinfer-ai/flashinfer/pull/5060 — Variant A: MXFP8 pre-attention for B200 (`sm_100a`) and B300 (`sm_103a`). Implementation and validation are complete; the PR is open for review.

## 3. Packed-varlen FlashAttention-4 kernels

Input after our existing Ulysses all-to-all:

```text
Q/K/V: [T, 56/P, 128]
cu_seqlens: int32
causal: false
output: BF16 [T, 56/P, 128]
```

Nominal `T` centers are 33,472, 38,592, 48,768, 58,944, 74,240, and
109,952. Requested standalone variants:

```text
A. BF16 FA4
B. E4M3 FP8 FA4 with explicit descales
C. experimental block-scaled NVFP4 FA4
```

For C, we would be interested in a genuine low-precision Tensor Core implementation
rather than NVFP4 storage followed by BF16 matmuls. Softmax remains FP32. It would
be helpful to report the actual QK and PV MMA dtypes separately; a hybrid
NVFP4-QK / FP8-or-BF16-PV candidate would also be useful.

## 4. Direct-layout attention output projection

Reference operation:

```text
inverse-Ulysses receive layout [P, M, 56/P, 128]
  -> read without a standalone unpack/transpose
  -> O GEMM [M, 7168] x [7168, 5376]
  -> indexed gate * output + residual
  -> BF16 [M, 5376]
```

Requested GEMM variants: BF16, MXFP8/FP8, and NVFP4 W4A4. The collective itself
is outside the kernel.

## 5. Fused RMSNorm + AdaLN + FC1 + SwiGLU

Reference operation:

```text
x [M, 5376]
  -> RMSNorm
  -> indexed AdaLN scale/shift
  -> FC1 [M, 5376] x [5376, 28672]
  -> split gate/up
  -> SiLU(gate) * up
  -> BF16 [M, 14336]
```

Requested GEMM variants: BF16, MXFP8/FP8, and NVFP4 W4A4.

## 6. Persistent full MLP candidate

Reference operation:

```text
RMSNorm + indexed AdaLN
  -> FC1 [M, 5376] x [5376, 28672]
  -> SwiGLU
  -> FC2 [M, 14336] x [14336, 5376]
  -> indexed gate * output + residual
  -> BF16 [M, 5376]
```

Requested variants: BF16, MXFP8/FP8, and NVFP4 W4A4. This is only worth keeping
if it beats kernel 5 plus a tuned standalone FC2 GEMM; CAKE does not need to solve
that scheduling decision outside the kernel benchmark.

## 7. Quantize-and-pack helper kernels

These are useful fallbacks when the GEMM epilogue cannot write the communication
layout directly:

```text
BF16 Q/K/V [M, 56, 128]
  -> destination-major E4M3 + scales [P, M, 56/P, 3, 128]

BF16 Q/K/V [M, 56, 128]
  -> destination-major NVFP4 + block scales [P, M, 56/P, 3, 128]
```

Quantization and relayout should happen in one HBM pass.

## RTX 5090 / SM120 quantized SP1 candidates

If CAKE currently supports SM120, we would also appreciate considering a smaller
set of single-GPU kernels for the quantized H3 FL2VA path on RTX 5090. This target
uses SP1 only, so there is no Ulysses layout or collective in these kernels.

For SP1, `M = T`; the nominal M values are 33,472, 38,592, 48,768, 58,944,
74,240, and 109,952 for 4/5/6/8/10/15 seconds. We are interested in two quantized
GEMM formats:

- W8A8 E4M3 FP8;
- NVIDIA-compatible block-scaled NVFP4 W4A4.

BF16 is used as the numerical reference but does not need a separately tuned 5090
kernel in this request.

### 5090-K1. Quantized fused pre-attention

```text
x [M, 5376]
  -> RMSNorm + indexed AdaLN
  -> FP8 or NVFP4 QKV GEMM [M, 5376] x [5376, 21504]
  -> BF16 Q/K RMSNorm + 3-D RoPE
  -> Q/K/V [M, 56, 128]
```

Useful output variants would be BF16, E4M3 plus scales, and NVFP4 plus block
scales, so that we can connect the best one to the corresponding attention path.

### 5090-K2. Quantized packed-varlen FA4

```text
Q/K/V [T, 56, 128]
cu_seqlens: int32
causal: false
output: BF16 [T, 56, 128]
```

The requested candidates are E4M3 FP8 FA4 and experimental block-scaled NVFP4
FA4. Softmax should remain FP32. For the NVFP4 candidate, it would be helpful to
report the QK and PV MMA dtypes separately and include a hybrid NVFP4-QK /
FP8-or-BF16-PV result if it is more accurate or faster.

### 5090-K3. Quantized attention output projection

```text
attention output [M, 7168]
  -> FP8 or NVFP4 O GEMM [M, 7168] x [7168, 5376]
  -> indexed gate * output + residual
  -> BF16 [M, 5376]
```

### 5090-K4. Quantized FC1 + SwiGLU

```text
x [M, 5376]
  -> RMSNorm + indexed AdaLN
  -> FP8 or NVFP4 FC1 [M, 5376] x [5376, 28672]
  -> SwiGLU
  -> BF16 [M, 14336]
```

### 5090-K5. Quantized persistent full MLP candidate

```text
RMSNorm + indexed AdaLN
  -> FP8 or NVFP4 FC1
  -> SwiGLU
  -> FP8 or NVFP4 FC2 [M, 14336] x [14336, 5376]
  -> indexed gate * output + residual
  -> BF16 [M, 5376]
```

As with the B300 candidates, this full MLP would only be useful if it outperforms
the fused FC1 + SwiGLU kernel followed by a tuned standalone FC2 kernel. We will
provide prepacked quantized weights and scales; weight conversion and calibration
are not part of the kernel request.

For RTX 5090, our preferred starting point would be **5090-K4 NVFP4 FC1 +
SwiGLU**, followed by 5090-K1 and the FP8/NVFP4 attention candidates.

## Integration

For any selected candidate, a FlashInfer-compatible kernel implementation and
basic benchmark results would be very helpful. We will handle the Sol
Engine/SGLang integration and end-to-end evaluation.

**Comment by yyihuang** [comment]
@Kim2026-dev Thank you for proposing this detailed CAKE kernel request! We have moved it into this sub-issue so we can track it independently.

**Comment by Edenzzzz** [comment]
hi @Kim2026-dev I'm the author of the nvfp4 attn on B300. I would not recommend quantization all of QKV to NVFP4 on B300, as group-wise P quantization stalls softmax significantly, unless we want to dequantize V before GEMM like Inkling does. It's better to quantize QK to NVFP4 and V to FP8, then pack it with uint8 `[P, M, H/P, 2D]`. 
I'm also working on an agent for kernels project and will try some of these tickets soon.

<img width="1800" height="750" alt="Image" src="https://github.com/user-attachments/assets/8e2821bc-58a5-411f-91d4-0482703f299b" />

**Comment by LiquidMovz** [comment]
@yyihuang @Kim2026-dev — Chase (ADAPT). We write custom Blackwell kernels (B200, B300, RTX PRO 6000). We can take a first kernel off this list.

Reading the issue as current; please correct anything that moved:

- Preferred start you named: **5090-K4 NVFP4 FC1+SwiGLU**, then 5090-K1, then the FP8/NVFP4 attention candidates
- B300 / SM103a list still live (Ulysses SP1/2/4/8)
- @Edenzzzz note: do not quantize all QKV to NVFP4 on B300; QK NVFP4 + V FP8, pack `uint8 [P, M, H/P, 2D]`

To write 5090-K4 we still need from you:

1. Confirm 5090-K4 NVFP4 is still first after that QK/V note
2. CUDA / driver / FlashInfer / SGLang / Sol Engine pins
3. The PyTorch reference + tensor traces you offered, for two M buckets (6s and 10s is enough)
4. Prepacked NVFP4 FC1 weights + the scale layout
5. Current us on 5090 for that op, and the accuracy gate vs BF16
6. When you want a benchable cubin
7. Off-thread contact if you prefer that to the issue
8. Who signs off on the list (and who Kim2026-dev is), if that is not you

We write the kernel. You keep Sol Engine / SGLang integration.
